### ISSUE TITLE
Old structures and entities remain on dungeon regen

### ISSUE THREAD

**Issue by MikeyCarter** [orphu/mcdungeon#298]
I have the fill_caves: true but don't think it's covering previous mcdungeons regenerations.   

This is creating a few situations.  We had one room which was a lava pit... but you could still see part of the old room in the walls.   Also had a case where previous regen had lava underneath a wood floor.  Thus the entire room went up in flames as soon as the chunk was loaded.

If the entire MC dungeon area was filled in with stone, or air before it started generating it would fix this.

**Comment by JiFish** [comment]
This would totally break the cavern room type and make ConstructionArea features somewhat less fun.

Rooms are not guaranteed to be stable. There are quite a few situations where MCDungeon itself generates combinations that cause the room to go up in flames. LavaChasm and any flammable floor for example.

I'm pretty sure we don't consider this a flaw. "Room that has/is gone up in flames" is just one of the things you can encounter...

> I have the fill_caves: true but don't think it's covering previous mcdungeons regenerations.

I'm pretty sure this only covers caves that are nearby, and maybe caves under the dungeon.

**Comment by MikeyCarter** [comment]
Problem with not clearing the area on regen is those cavern/construction areas could be filled with command blocks and partial redstone circuts.   Rather than filling with stone... any way of filling with more random or the seeded area?

**Comment by JiFish** [comment]
> Problem with not clearing the area on regen is those cavern/construction areas could be filled with command blocks and partial redstone circuts.

Is that not accounted for?

**Comment by orphu** [comment]
I'm not sure about the old parts still lying around... One of the things regen is supposed to do  is fill in the area of the dungeon with fake stone and ore distribution and remove all entities. This should completely wipe out any old rooms. 

You can see this here:  https://github.com/orphu/mcdungeon/blob/development/dungeon.py#L2209

So for clarity, you are generating dungeons without fill_caves, and then trying to use it on regen? 

If so, this is probably points out something really dangerous and should not be allowed. fill_caves assumes the chunk finder already took structures into account. Adding it later will almost certainly destroy things in the surrounding area.

**Comment by orphu** [comment]
Ahh, wait, I think fill_caves already takes this into account. If a chunk is not a "good chunk" ie: it has an identified structure in it, it won't be filled in. So, adding fill_caves later will probably result in areas not being filled in.

**Comment by MikeyCarter** [comment]
I'm using the same config to regen as I did to initially create the dungeon.   So fill_caves: true is set for both cases.

**Comment by orphu** [comment]
Okay, I'll test this out. Sounds like it's not working as intended.

**Comment by MikeyCarter** [comment]
Related... Couldn't figure out why I had these random villagers wandering around... then realized after a regen any mobs are probably staying where they are.   Not that that's a bad thing..  zombies will kill them off eventually.

**Comment by JiFish** [comment]
Hmmm... we should probably remove entities. Things like armoury champions shouldn't just be wandering around.

**Comment by MikeyCarter** [comment]
![2015-04-20_18 21 13](https://cloud.githubusercontent.com/assets/1325499/7241812/c0c8ea4c-e78a-11e4-8dc5-05143aca8c3c.png)

Just caught an image of the sort of weirdness I'm seeing.

**Comment by orphu** [comment]
That is very odd, and certainly not normal. I've not tried to replicate yet.

**Comment by orphu** [comment]
Yup. This is a bug with regen and dungeons generated with fill_caves. 

It *should* remove tile/mob entites and fill in with resources but that part is not taking into account the dungeon offsets for the large size.

**Comment by orphu** [comment]
Hmmm, I'm actually having a hard time reproducing this. 

I've tried generating with and without fill_caves and then regenerating with and without fill_caves and it appears to be working correctly. 

Can you provide step by step how you did this? What version did you use to generate the dungeon originally?

**Comment by MikeyCarter** [comment]
Was originally generated as 0.14.1 (at least that's what's on the treasure maps generated at the same time).   It has since been regenerate many many times.   I usually do a regen after at least every friday before "minecraft" day.   But some times I'll regen a few times for other reasons.

Only started noticing the weird stuff in the last two weeks.

**Comment by orphu** [comment]
Can you pastebit/gist the entire regen output? 

Can you somehow send me your map?

**Comment by MikeyCarter** [comment]
As requested:

http://pastebin.com/download.php?i=nWyWkisJ

https://drive.google.com/open?id=0B9i4aAtBXfTVWFprVmpTZUs2WXM&authuser=0

Hopefully you can access the map via google drive... if not I'll try something else.

**Comment by JiFish** [comment]
Just a note that someone else posted about this on the forum thread. Unfortunately, they didn't provide further details.

**Comment by orphu** [comment]
I'll be curious to see if this is still a problem in 0.15/1.9. My initial testing hasn't been able to reproduce the issue.

**Comment by orphu** [comment]
I've tested regenerating different types of dungeons with and without fill_caves under the 0.15/1.9 branch and I can't seem to repro this. I'm wondering if it was a problem that was fixed in utils.py having to do with reading the dungeon info. Things got a bit weird in 1.8 with the item ID transition, and maybe something was going on there. Now that everything is string IDs, this may be fixed.  

I'm going to remove this from the 0.15 milestone and close it for now. If it comes up again I'll take a look.
